### PR TITLE
fall back to slower stat filesize if optimized filesize fails

### PR DESCRIPTION
**Pull request by IanButterworth**
Based on the triage suggestion https://github.com/JuliaLang/julia/issues/54203#issuecomment-2102780379

### CODE REVIEW

**Comment by giordano** [comment]
@ronisbr would you be able to test this?

File: base/iostream.jl [hunk 1]
```diff
@@ -228,7 +228,13 @@ function position(s::IOStream)
...
+    sz = try
+        @_lock_ios s ccall(:ios_filesize, Int64, (Ptr{Cvoid},), s.ios)
+    catch e
+        e isa IOError || rethrow()
+        # if `s` is not seekable `ios_filesize` can fail, so fall back to slower stat method
+        filesize(stat(s))
+    end
     if sz == -1
         err = Libc.errno()
         throw(IOError(string("filesize: ", Libc.strerror(err), " for ", s.name), err))
```
Comment (vtjnash) [inline base/iostream.jl:240]:
```suggestion
    sz = @_lock_ios s ccall(:ios_filesize, Int64, (Ptr{Cvoid},), s.ios)
    if sz == -1
        # if `s` is not seekable `ios_filesize` can fail, so fall back to slower stat method
        sz = filesize(stat(s))
```

**Comment by ronisbr** [comment]
> @ronisbr would you be able to test this?

I will try this weekend! I am currently without access to a Raspberry Pi to test. 

@JakeZw

Can you test this PR?

**Comment by KristofferC** [comment]
Or copy paste the new function in a `@eval Base` block.

**Comment by JakeZw** [comment]
I used giordano's method.

The command that has been causing problems in BaremetalPi is gpio_init(), in which the following commands are executed and then stored in a structure.
```
        using Mmap
        gpiomem_io  = open("/dev/gpiomem", "w+");
        gpiomem_map = Mmap.mmap(
            gpiomem_io,
            Vector{UInt32},
            (1024,),
            0,
            grow = false
        );

```

With Julia 1.10.0 it all works.

On the development version that I just downloaded
I get the following output
```
julia> gpiomem_map = Mmap.mmap(
           gpiomem_io,
           Vector{UInt32},
           (1024,),
           0,
           grow = false
       );
ERROR: ArgumentError: requested size 4096 larger than file size 0, but requested not to grow
Stacktrace:
 [1] mmap(io::IOStream, ::Type{Vector{UInt32}}, dims::Tuple{Int64}, offset::Int64; grow::Bool, shared::Bool)
   @ Mmap ~/Downloads/julia-984f884dfe/share/julia/stdlib/v1.12/Mmap/src/Mmap.jl:226
 [2] top-level scope
   @ REPL[22]:1
```
Awaiting instructions.

**Comment by JakeZw** [comment]
> Or copy paste the new function in a `@eval Base` block.

This is a bit cryptic for me.

**Comment by JakeZw** [comment]
Still not working

```
julia> gpiomem_map = Mmap.mmap(
           gpiomem_io,
           Vector{UInt32},
           (1024,),
           0,
           grow = true
       );
ERROR: SystemError: ftruncate: Invalid argument
Stacktrace:
 [1] systemerror(p::Symbol, errno::Int32; extrainfo::Nothing)
   @ Base ./error.jl:185
 [2] systemerror
   @ ./error.jl:184 [inlined]
 [3] grow!(io::IOStream, offset::Int64, len::Int64)
   @ Mmap ~/Downloads/julia-984f884dfe/share/julia/stdlib/v1.12/Mmap/src/Mmap.jl:91
 [4] mmap(io::IOStream, ::Type{Vector{UInt32}}, dims::Tuple{Int64}, offset::Int64; grow::Bool, shared::Bool)
   @ Mmap ~/Downloads/julia-984f884dfe/share/julia/stdlib/v1.12/Mmap/src/Mmap.jl:224
 [5] top-level scope
   @ REPL[24]:1

julia>
```

**Comment by JakeZw** [comment]
I don't think filesize() is the source of the BaremetalPi problem.  I compared the working and not working versions of the function filesize and they don't correspond to when the problem occurs.  See the comments below for what is working and what is not working.  I also took at look Mmap.map and it has changed from 1.10.0 to 1.10.1 and is different again for 1.12.0.  I will report back to the BaremetalPi issue tracker.

```

# version 1.10.0  BaremetalPi works
function filesize(s::IOStream)
sz = @_lock_ios s ccall(:ios_filesize, Int64, (Ptr{Cvoid},), s.ios)
if sz == -1
err = Libc.errno()
throw(IOError(string("filesize: ", Libc.strerror(err), " for ", s.name), err))
end
return sz
end

version 1.10.5 BaremetalPi does not work
function filesize(s::IOStream)
sz = @_lock_ios s ccall(:ios_filesize, Int64, (Ptr{Cvoid},), s.ios)
if sz == -1
err = Libc.errno()
throw(IOError(string("filesize: ", Libc.strerror(err), " for ", s.name), err))
end
return sz
end

# version 1.12.0 BaremetalPi does not work
function filesize(s::IOStream)
sz = @_lock_ios s ccall(:ios_filesize, Int64, (Ptr{Cvoid},), s.ios)
if sz == -1
# if `s` is not seekable `ios_filesize` can fail, so fall back to slower stat method
sz = filesize(stat(s))
end
return sz
end
```

**Comment by ronisbr** [comment]
The problem is the `filesize` function. The modification `Mmap.mmap` now calls the function `filesize` (which was not being called before) that leads to the problem in raspberry pi due to the file `/dev/gpiomem`.

**Comment by JakeZw** [comment]
Filesize(io) returns 0 with this patch so it is working.  

I see what you mean about Mmap,mmap having been changed from 1.10.0 to 1.10.1.  Unfortunately it has changed again for 1.12.0, so that may be a new issue.

**Comment by vtjnash** [comment]
I think we just need a check for `isfile(t)` on the stat result before using the result of the `filesize`, if the result is `0`, since that filters out non-regular files, since we know that will fail the ftruncate syscall, but also won't be useful to call:

>       EINVAL fd does not reference a regular file or a POSIX shared memory object.

**Comment by vtjnash** [comment]
(in addition to this PR or after this PR)

**Comment by JakeZw** [comment]
It is failing on line 91 of grow! in the Mmap.jl file where the systemcall on line 90 returns failure = -1.  Complete error message shown above.

**Comment by IanButterworth** [comment]
@vtjnash I gave that a go here. Could be separated into another PR, but for testing purposes here for now. 

I couldn't see a tidy way to reuse the `stat` call within `filesize`

**Comment by JakeZw** [comment]
if I change your line to 
```
filelen == 0 && !isfile(io) && return  # note the ! before isfile
```
I no longer get an error.  I am in process of testing this with BaremetalPi.

It seems that @ronisbr prevented Mmap from using grow! by setting the option grow=false when mmap was called.  With this fix I need to set grow=true, but then grow! returns before performing any function, which seems counterintuitive.  I am wondering if the following logic in mmap should be reviewed and the case of !isfile(io) included in the logic there?
```
    @static if Sys.isunix()
        prot, flags, iswrite = settings(file_desc, shared)
        if requestedSizeLarger
            if iswrite
                if grow
                    grow!(io, offset, len)
                else
                    throw(ArgumentError("requested size $szfile larger than file size $(filesize(io)), but requested not to grow"))
                end
            else
                throw(ArgumentError("unable to increase file size to $szfile due to read-only permissions"))
            end
        end
```

**Comment by ronisbr** [comment]
> It seems that @ronisbr prevented Mmap from using grow! by setting the option grow=false when mmap was called. With the current proposed fix I need to set grow=true, but then grow! returns before performing any function, which seems counterintuitive. I am wondering if the following logic in mmap (lines 220 to 231) should be reviewed and the case of !isfile(io) included in the logic there?

Yes, I agree. However, I can change BaremetalPi.jl to use `grow = true` if required. It is just that technically you do not need to grow that file.

**Comment by JakeZw** [comment]
With the fix shown two posts up and a dev version of BaremetalPi with grow set to true, my application that uses BaremetalPi is working.

**Comment by JakeZw** [comment]
I don't pretend to understand what Mmap is doing, but I took another look to see if I could make it not call grow! in our case.  I put an if statement as shown with code starting at line 214:
```
Mmap.jl starting at line 214
if !(io isa Mmap.Anonymous)
    @static if !Sys.isapple()
    if isfile(io)   # added this statement
        requestedSizeLarger = szfile > filesize(io)
    end
end
end
```
This prevents the grow statement from being called in my use case.
This would definitely need a set of eyes that understands what is going on better.  However with my use case with the raspberry pi it is working.

**Comment by JakeZw** [comment]
Probably better than the above modification is the following change on line 222 where the code block below starts on line 211.  This will only execute grow! if io is a file.  I tested this with the keyword argument grow=false, and with the unmodified version of BaremetalPi and both worked.
```
file_desc = gethandle(io)
szfile = convert(Csize_t, len + offset)
requestedSizeLarger = false
if !(io isa Mmap.Anonymous)
    @static if !Sys.isapple()
requestedSizeLarger = szfile > filesize(io)
end
end
# platform-specific mmapping
@static if Sys.isunix()
    prot, flags, iswrite = settings(file_desc, shared)
    if isfile(io) && requestedSizeLarger  # add a condition to this line to ensure it only checks files
        if iswrite
            if grow
                grow!(io, offset, len)
           else
               throw(ArgumentError("requested size $szfile larger than file size $(filesize(io)), but requested not to grow"))
            end
        else
            throw(ArgumentError("unable to increase file size to $szfile due to read-only permissions"))
       end
    end

```

**Comment by IanButterworth** [comment]
Great. Feel free to open another PR with the content of this PR and your suggested changes

**Comment by JakeZw** [comment]
I feel free, but my PR knowledge is non-existent so I found this [tutorial](https://kshyatt.github.io/post/firstjuliapr/) and got stuck on the zeroth step!

**Comment by ronisbr** [comment]
@IanButterworth I will test this PR and, if it does not work, I create a new one with suggestions. I just need to have access to a RPI. If everything goes right, I will be able to do so in this weekend.

**Comment by JakeZw** [comment]
@ronisbr if you can make the PR I can test.  If it is identical to what I have done, it has been tested but feel free to ping me for any testing.

**Comment by ronisbr** [comment]
Sure! Thanks @JakeZw !

**Comment by ronisbr** [comment]
Just an update, the RPI was not delivered yet :( I will make the PR as soon as I receive it.

**Comment by ronisbr** [comment]
Hi @IanButterworth !

Sorry for the delay, for some reason Julia 1.12 is **extremely** slow in Raspberry Pi Zero 2, and I had a lot of crashes trying to use the development version. I took me sometime to be able to install BaremetalPi.jl and test the modification.

Regarding your PR, I think there is only one mistake:

The line:

```julia
filelen == 0 && isfile(stat(io)) && return
```

should be

```julia
filelen == 0 && !isfile(stat(io)) && return
```

as @JakeZw mentioned before. From all modifications, this seems to me the best one. Indeed, it seems reasonable to me: do not try to grow the file if it has 0 length and it is not a regular file. It is better than what was proposed latter by @JakeZw, which will skip grow if the file is not a regular file. Maybe there will be some regressions in corner cases?

After those modifications, I could load the file using Mmap, and BaremetalPi.jl will work after I remove `grow = false`.

Question: since this is a very minor change to your PR, can you update it please? If not, I can make a new one.

File: stdlib/Mmap/src/Mmap.jl [hunk 2]
```diff
@@ -86,6 +86,8 @@ grow!(::Anonymous,o::Integer,l::Integer) = return
 function grow!(io::IO, offset::Integer, len::Integer)
     pos = position(io)
     filelen = filesize(io)
+    # If non-regular file skip trying to grow since we know that will fail the ftruncate syscall
+    filelen == 0 && isfile(stat(io)) && return
```
Comment (oscardssmith) [inline stdlib/Mmap/src/Mmap.jl:90]:
```suggestion
    filelen == 0 && isfile(stat(io)) && return
```

**Comment by oscardssmith** [comment]
change made. let's merge this.

**Comment by IanButterworth** [comment]
Sounds good but this needs a test.

**Comment by JakeZw** [comment]
I have tested all the modification that I have proposed with a Raspberry Pi 4 and they worked.  Are there any other tests you would like me to perform?

**Comment by oscardssmith** [comment]
by "needs a test", Ian means it needs a test in CI, but it's somewhat unclear how we can do that.

**Comment by IanButterworth** [comment]
I was imagining that we could either use some address on a linux system to mimic the gpio output, or mock it.
If that isn't easy then I guess merge without tests, but open an issue to note that it's untested?

**Comment by JakeZw** [comment]
From https://github.com/ronisbr/BaremetalPi.jl/blob/master/src/gpio.jl

the commands that errored are:

`        gpiomem_io  = open("/dev/gpiomem", "w+")
        gpiomem_map = Mmap.mmap(
            gpiomem_io,
            Vector{UInt32},
            (GPIOMEM_SIZE,),
            0,
            grow = false
        )`

If we change grow = true the proposed changes work.

If we leave grow = false then we need the additional change on line 222 that I proposed above.

This code segment should work as a test.

**Comment by ronisbr** [comment]
The problem is that this scenario only works in a Raspberry Pi, which cannot be emulated by CI here in GitHub.

File: stdlib/Mmap/src/Mmap.jl [hunk 3]
```diff
@@ -86,6 +86,8 @@ grow!(::Anonymous,o::Integer,l::Integer) = return
 function grow!(io::IO, offset::Integer, len::Integer)
     pos = position(io)
     filelen = filesize(io)
+    # If non-regular file skip trying to grow since we know that will fail the ftruncate syscall
+    filelen == 0 && !isfile(stat(io)) && return
```
Comment (IanButterworth) [inline stdlib/Mmap/src/Mmap.jl:90]:
```suggestion
```

**Review by IanButterworth — COMMENTED** [review]
Given there are ongoing questions here about what is needed to fix Mmap, I will reduce this to just the issue in the title so it can be merged, then another PR opened to fix the remaining issues.

**Comment by IanButterworth** [comment]
Also, seems we don't have a good idea for a test here, so I'll remove that needs label.

**Comment by ronisbr** [comment]
Hi @IanButterworth 

> Given there are ongoing questions here about what is needed to fix Mmap, I will reduce this to just the issue in the title so it can be merged, then another PR opened to fix the remaining issues.

Is it really necessary? I thought everything was fine with that suggestion. The current state does not fix #54203. Since this is an important regression (nobody is able to use Julia in the current state to access Raspberry Pi's GPIO), can't we try to implement the solution to #54203 in this PR to fix the problem faster?

**Comment by ronisbr** [comment]
By the way, the problem is that this PR:

https://github.com/JuliaLang/julia/pull/44354/files

now perform checks that were not performed before (like throwing an error in some conditions).

**Comment by IanButterworth** [comment]
The Mmap fixes have been opened here https://github.com/JuliaLang/julia/pull/55849

@JakeZw opening PRs isn't as complicated as the guide at https://github.com/JuliaLang/julia/pull/55641#issuecomment-2334347125 makes out. 

@ronisbr there's just been a lot of noise in this PR and the title fix should get in. See https://github.com/JuliaLang/julia/pull/55849 for the remainder. Please review & approve that if it is correct.

**Comment by ronisbr** [comment]
> @ronisbr there's just been a lot of noise in this PR and the title fix should get in. See https://github.com/JuliaLang/julia/pull/55849 for the remainder. Please review & approve that if it is correct.

Perfect @IanButterworth ! You were very fast :) I just tested both modifications and, from my side, everything is working perfectly (Raspberry Pi Zero 2). However, I think it will be important if @JakeZw test is too.

**Pull request merged**